fastai/text/data.py

```diff
@@ -13,44 +13,106 @@
 text_extensions = {'.txt'}
 
 class LanguageModelPreLoader(Callback):
-    "Transforms the texts in `dataset` in a stream for language modelling."
-    def __init__(self, dataset:LabelList, lengths:Collection[int]=None, bs:int=64, bptt:int=70, backwards:bool=False, 
-                 shuffle:bool=False, drop_last:bool=False):
-        self.dataset,self.bs,self.bptt,self.backwards = dataset,bs,bptt,backwards
-        self.shuffle,self.drop_last = shuffle,drop_last
-        self.lengths = np.array(ifnone(lengths, [len(o) for o in dataset.x.items]))
-        self.n = self.lengths.sum() // self.bs
-        
-    def __len__(self):
-        if self.item is not None:    return 1
-        if (self.n-1)%self.bptt == 0 or self.drop_last: return ((self.n-1) // self.bptt) * self.bs
-        return ((self.n-1) // self.bptt + 1) * self.bs
+    "Create a dataloader with bptt slightly changing."
     
+    class CircularIndex():
+        #When the index exceeds the length of self.idx then it is wrap to start at the head 
+        #or the end if indexing is moving backwards. The shuffled is done in-place, 
+        def __init__(self, length:int, forward:bool): 
+            self.idx      = np.arange(length)
+            self.forward_ = forward
+        def __getitem__(self, i): 
+            idx = self.idx
+            return idx[ i%len(idx) if self.forward_ else len(idx)-1-i%len(idx) ]
+        def __len__(self) -> int: return len(self.idx)
+        def shuffle(self): np.random.shuffle(self.idx)
+        def forward(self, forward:bool=True): self.forward_ = forward
+
+    def __init__(self, dataset:LabelList, bs:int=32, bptt:int=70, backwards:bool=False, shuffle:bool=False):
+        self.dataset,self.bs,self.bptt,self.shuffle,self.backwards = dataset,bs,bptt,shuffle,backwards
+        self.totalToks = 0
+        for rag in dataset.x.items: self.totalToks += len(rag)
+        self.ite_len   = self.bs*int( math.ceil( self.totalToks/(self.bptt*self.bs) )) if self.item is None else 1
+        self.idx = None
+
+    def __len__(self): return self.ite_len
     def __getattr__(self,k:str)->Any: return getattr(self.dataset, k)
-    
+   
+    def allocate_buffers(self):     
+        "allocate the required worth-case batchbuffer"
+        #idx:     used to index into items so that shuffle is used - if idx has been shuffled 
```

Comment:
This is all for the documentation of fill_forward I believe, and has nothing to do here.
Can we have better names for ei and eo?

Reply:
what about ro and ri
ro being the offset where each batch row starts and ri being the index into the current rag